browser/brave_news/brave_news_tab_helper.cc

```diff
@@ -89,48 +102,89 @@ bool BraveNewsTabHelper::IsSubscribed(const FeedDetails& feed_details) {
 }
 
 bool BraveNewsTabHelper::IsSubscribed() {
-  for (const auto& feed : GetAvailableFeeds()) {
-    if (IsSubscribed(feed))
+  for (const auto& feed_url : GetAvailableFeedUrls()) {
+    if (IsSubscribed(feed_url)) {
       return true;
+    }
   }
   return false;
 }
 
-void BraveNewsTabHelper::ToggleSubscription(const FeedDetails& feed_details) {
-  bool subscribed = IsSubscribed(feed_details);
-  auto* publisher = controller_->publisher_controller()->GetPublisherForFeed(
-      feed_details.feed_url);
+std::string BraveNewsTabHelper::GetTitleForFeedUrl(const GURL& feed_url) {
+  auto* default_publisher =
+      controller_->publisher_controller()->GetPublisherForFeed(feed_url);
+  if (default_publisher) {
+    return default_publisher->publisher_name;
+  }
+
+  auto it = base::ranges::find_if(
+      rss_page_feeds_,
+      [feed_url](const auto& details) { return details.feed_url == feed_url; });
+  if (it == rss_page_feeds_.end()) {
+    return "";
+  }
+
+  if (it->title.empty()) {
+    const auto url = web_contents()->GetLastCommittedURL();
+    controller_->FindFeeds(
+        feed_url,
+        base::BindOnce(&BraveNewsTabHelper::OnFoundFeeds,
+                       weak_ptr_factory_.GetWeakPtr(), feed_url, url));
+    return feed_url.spec();
+  }
+  return it->title;
+}
+
+void BraveNewsTabHelper::ToggleSubscription(const GURL& feed_url) {
+  bool subscribed = IsSubscribed(feed_url);
+  auto* publisher =
+      controller_->publisher_controller()->GetPublisherForFeed(feed_url);
   if (publisher) {
     controller_->SetPublisherPref(
         publisher->publisher_id, subscribed
                                      ? brave_news::mojom::UserEnabled::DISABLED
                                      : brave_news::mojom::UserEnabled::ENABLED);
   } else if (!subscribed) {
-    controller_->SubscribeToNewDirectFeed(feed_details.feed_url,
-                                          base::DoNothing());
+    controller_->SubscribeToNewDirectFeed(feed_url, base::DoNothing());
   }
 }
 
 void BraveNewsTabHelper::OnReceivedRssUrls(const GURL& site_url,
                                            std::vector<GURL> feed_urls) {
-  if (site_url != GetWebContents().GetLastCommittedURL())
+  if (site_url != GetWebContents().GetLastCommittedURL()) {
     return;
+  }
 
+  // First things first, we just store the urls. This let's us know that we have
+  // feeds, so we should show the button.
   for (const auto& url : feed_urls) {
-    controller_->FindFeeds(
-        url, base::BindOnce(&BraveNewsTabHelper::OnFoundFeeds,
-                            weak_ptr_factory_.GetWeakPtr(), site_url));
+    rss_page_feeds_.push_back({.feed_url = url, .title = ""});
   }
+
+  AvailableFeedsChanged();
 }
 
 void BraveNewsTabHelper::OnFoundFeeds(
```

Comment:
Should the name of this function be something more like OnFoundFeedData since this is only ever used for 1 specific feed now, and not a whole page's worth of feeds? Am I even understanding that correctly?